Update purldb-toolkit and related README files #249 (#450)
nexB/purldb: purldb-toolkit/README.rst

purldb-toolkit
==============

.. contents:: :local:
    :depth: 3

> 

Reviewer: One blank line is enough here, instead of 3.

Author: Done.

purldb-toolkit is command line utility and library to use the PurlDB, its API and various related libraries.

The ``purlcli`` command acts as a client to the PurlDB REST API end point(s) to expose PURL services.
It serves both as a tool, as a library and as an example on how to use the services programmatically.

> 

Reviewer: `pip install purldb-toolkit` should be in a console section

Author: Done -- wrapped `pip install purldb-toolkit` in a `.. code-block:: console` directive.

    $ purlcli --help
    Usage: purlcli [OPTIONS] COMMAND [ARGS]...

      Return information from a PURL.

    Options:
      --help  Show this message and exit.

    Commands:

>       metadata  Given one or more PURLs, for each PURL, return a mapping of...

Reviewer: Here instead of three dots at the end, we can show the full texts which are there when we run `purlcli --help`, for all 4 commands here. Otherwise the truncated texts are not that useful.

Author: Actually @AyanSinhaMahapatra when I ran `purlcli --help` I got the same truncated texts.  I'll postpone this change until you've had a chance to read this reply in case that changes your thinking.  I like the idea of more info, but as you'll see in the examples below, the description for instance for `metadata` is rather long -- just want to make sure we're OK with the way this wraps due to its length when we insert it into the `purlcli --help` output block in the RTD.

This is `purlcli --help`:

```
$ purlcli --help
Usage: purlcli [OPTIONS] COMMAND [ARGS]...

  Return information from a PURL.

Options:
  --help  Show this message and exit.

Commands:
  metadata  Given one or more PURLs, for each PURL, return a mapping of...
  urls      Given one or more PURLs, for each PURL, return a list of all...
  validate  Check the syntax and upstream repo status of one or more PURLs.
  versions  Given one or more PURLs, return a list of all known versions...
```

And this is `purlcli metadata --help`:

```
$ purlcli metadata --help
Usage: purlcli metadata [OPTIONS]

  Given one or more PURLs, for each PURL, return a mapping of metadata fetched
  from the fetchcode package.py info() function.

Options:
  --purl TEXT        PackageURL or PURL.
  --output FILENAME  Write meta output as JSON to FILE.  [required]
  --file FILENAME    Read a list of PURLs from a FILE, one per line.
  --help             Show this message and exit.
```

      urls      Given one or more PURLs, for each PURL, return a list of all...
      validate  Check the syntax and upstream repo status of one or more PURLs.
      versions  Given one or more PURLs, return a list of all known versions...

And the following subcommands:

> 'validate' -- validate a PURL

Reviewer: ```suggestion
``validate`` : validate a PURL
```
Looks better this way. It's better to wrap command names in double ` to highlight

Reviewer: Consider the same for all other headings

Author: Nice improvement -- done.

                        "pkg:nginx/nginx@0.8.9"
                    ],
                    "--file": null,
                    "--output": "<path/to/output.json>"
                },
                "errors": [],
                "warnings": [
                    "'check_existence' is not supported for 'pkg:nginx/nginx@0.8.9'"
                ]
            }
        ],
        "packages": [
            {
                "purl": "pkg:npm/canonical-path@1.0.0",
                "valid": true,
                "exists": true,
                "message": "The provided Package URL is valid, and the package exists in the upstream repo."
            },
            {
                "purl": "pkg:nginx/nginx@0.8.9",
                "valid": true,
                "exists": null,
                "message": "The provided PackageURL is valid, but `check_existence` is not supported for this package type."
            }
        ]
    }

**Submit multiple PURLs using a .txt file:**

.. code-block:: none

    purlcli validate --file <path/to/output.txt> --output <path/to/output.json>

*Sample input.txt:*

.. code-block:: console

    pkg:npm/canonical-path@1.0.0
    pkg:nginx/nginx@0.8.9

> Notes

Reviewer: Maybe instead of notes, we call this `Details` and give a link to the related API page/other docs as applicable (we might not have these API pages yet, feel free to create these too, see https://github.com/nexB/purldb/issues/445#issuecomment-2140684111)

Reviewer: Consider the same for all the similar sections in commands below.

Author: Heading changed to `Details` for all four commands.  

I've updated the `validate` sentence as you suggested.  

For the `versions` and `metadata` commands (which rely on fetchcode's `package_versions.py` and `package.py`, respectively), I removed the links to the actual code as you requested, but as far as I can tell there are no documents or similar sources to link to atm.  This is also the case for the `urls` command (which calls packageurl-python's `purl2url.py`).  As a result we currently have a `Details` link only for the `validate` command.

#######

> ``validate`` calls the ``public.purldb.io/api/validate/`` endpoint.

Reviewer: ```suggestion
``validate`` calls the ``validate/`` endpoint of the `purldb API <https://public.purldb.io/api/>`_
```

Author: Definitely better -- done.

----

'versions' -- collect package versions for a PURL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: none

    $ purlcli versions  --help
    Usage: purlcli versions [OPTIONS]

      Given one or more PURLs, return a list of all known versions for each PURL.

    Options:
      --purl TEXT        PackageURL or PURL.
      --output FILENAME  Write versions output as JSON to FILE.  [required]
      --file FILENAME    Read a list of PURLs from a FILE, one per line.
      --help             Show this message and exit.

Examples
########

**Submit multiple PURLs using the command line:**

.. code-block:: none

    purlcli versions --purl pkg:npm/canonical-path --purl pkg:nginx/nginx --output <path/to/output.json>

*Sample output:*

> .. code-block:: console

Reviewer: ```suggestion
.. code-block:: json
```

Reviewer: Consider the same for all applicable JSON snippets

Author: All blocks updated as needed to `.. code-block:: console`, `.. code-block:: json` or `.. code-block:: text`.

    {
        "headers": [
            {
                "tool_name": "purlcli",
                "tool_version": "0.2.0",
                "options": {
                    "command": "versions",
                    "--purl": [
                        "pkg:npm/canonical-path",
                        "pkg:nginx/nginx"
                    ],
                    "--file": null,
                    "--output": "<path/to/output.json>"
                },
                "errors": [],
                "warnings": [
                    "'pkg:nginx/nginx' not supported with `versions` command"
                ]
            }
        ],
        "packages": [
            {
                "purl": "pkg:npm/canonical-path@0.0.1",
                "version": "0.0.1",
                "release_date": "2013-12-19"
            },
            {
                "purl": "pkg:npm/canonical-path@0.0.2",
                "version": "0.0.2",
                "release_date": "2013-12-19"
            },
            {
                "purl": "pkg:npm/canonical-path@1.0.0",
                "version": "1.0.0",
                "release_date": "2018-10-24"
            }
        ]
    }

Notes
#######

> ``versions`` calls ``versions()`` from `fetchcode/package_versions.py <https://github.com/nexB/fetchcode/blob/master/src/fetchcode/package_versions.py>`__.

Reviewer: Maybe we need better documentation on fetchcode features and what ecosystems are supported (like at https://github.com/nexB/fetchcode/blob/master/src/fetchcode/package_versions.py#L39 but this is generated programmatically in the docs, see an example in https://github.com/nexB/scancode-toolkit/commit/67256a1d5312d2b9a09e99ca3f53f7bd3c658ccc) This should be in fetchcode docs and linked here. This is probably out of scope but something to consider for later

Reviewer: Basically we should avoid linking to code directly, as this is bad practice, and also the code might change. Easier to link to a section in another docs (via intersphinx)

Author: I removed the direct links to code as you requested -- I'd added these (for `versions` and `metadata`, which rely on `fetchcode/package_versions.py` and `fetchcode/package.py`, respectively) only because I could find no documentation or similar references to link to.

Version information is not needed in submitted PURLs and, if included, will be removed before processing.

----

'metadata' -- collect package metadata for a PURL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: console

    $ purlcli metadata --help
    Usage: purlcli metadata [OPTIONS]

      Given one or more PURLs, for each PURL, return a mapping of metadata fetched
      from the fetchcode package.py info() function.

    Options:
      --purl TEXT        PackageURL or PURL.
      --output FILENAME  Write meta output as JSON to FILE.  [required]
      --file FILENAME    Read a list of PURLs from a FILE, one per line.
      --help             Show this message and exit.

Examples
########

**Submit multiple PURLs using the command line:**

.. code-block:: none

    purlcli metadata --purl pkg:openssl/openssl@3.0.6 --purl pkg:nginx/nginx@0.8.9 --purl pkg:gnu/glibc@2.38 --output <path/to/output.json>

*Sample output:*

> .. code-block:: console

Reviewer: This should be consistent, JSON and console should be used consistently as applicable, for JSON snippets and commands respectively across this page.

Author: Good point and done.

                "download_url": "https://ftp.gnu.org/pub/gnu/glibc/glibc-2.38.tar.gz",
                "api_url": null,
                "size": null,
                "sha1": null,
                "md5": null,
                "sha256": null,
                "sha512": null,
                "bug_tracking_url": null,
                "code_view_url": null,
                "vcs_url": null,
                "copyright": null,
                "license_expression": null,
                "declared_license": null,
                "notice_text": null,
                "root_path": null,
                "dependencies": [],
                "contains_source_code": null,
                "source_packages": [],
                "repository_homepage_url": null,
                "repository_download_url": null,
                "api_data_url": null
            }
        ]
    }

Notes
#######

``metadata`` calls ``info()`` from `fetchcode/package.py <https://github.com/nexB/fetchcode/blob/master/src/fetchcode/package.py>`__.

The intended output for each PURL type supported by the ``metadata`` command is

- an input PURL with a version: output the metadata for the input version

> - an input PURL with no version: output a list of the metadata for all versions

Reviewer: ```suggestion
- an input PURL without a version: output a list of the metadata for all versions
```

                        "get_request_status_code": 200,
                        "head_request_status_code": 200
                    }
                ],
                "repository_download_url": {
                    "url": null,
                    "get_request_status_code": "N/A",
                    "head_request_status_code": "N/A"
                },
                "repository_homepage_url": {
                    "url": "https://rubygems.org/gems/rails/versions/7.0.0",
                    "get_request_status_code": 200,
                    "head_request_status_code": 200
                }
            }
        ]
    }

Notes
#######

- None atm.

Testing
-------

Run all purldb tests:

.. code-block:: none

    make test

Run all purlcli non-live tests (i.e., no live network calls):

.. code-block:: none

    DJANGO_SETTINGS_MODULE=purldb_project.settings pytest -vvs purldb-toolkit/tests/test_purlcli.py

Run all purlcli live tests (i.e., check actual API endpoints etc.)

.. code-block:: none

    DJANGO_SETTINGS_MODULE=purldb_project.settings pytest -vvs purldb-toolkit/tests/test_purlcli_live.py --run_live_fetch

> Funding

Reviewer: This should ideally be moved to a separate page under root, maybe under a misc folder.

Reviewer: Same for the LICENSE section below.

Author: @AyanSinhaMahapatra   I've created a new `Miscellaneous` folder under root containing `index.rst`, `funding.rst`, `license.rst` and `testing.rst` and moved the existing relevant text to those pages (for both PURLdb and PURLdb Toolkit).

BTW, in the RTD I see references to 

- `purldb-toolkit` (which I've been using as a placeholder name) and `PURLdb Toolkit`
- `PurlDB`, `PURLdb` and `purldb` (including `purldb is a trademark of nexB Inc.`)
- `purl` and `PURL`
- `ScanCode toolkit` and `scancode-toolkit`
- `Clearlydefined` and `ClearlyDefined`
- `Package matching`, `package matching`, `Package data`, `package data`, `package metadata`
- `MatchCode` and `Matchcode` and `matchcode`

We should probably decide on standard references.

Reviewer: > I've created a new Miscellaneous folder under root containing index.rst, funding.rst, license.rst and testing.rst and moved the existing relevant text to those pages (for both PURLdb and PURLdb Toolkit).

This makes sense. 

> We should probably decide on standard references.

Yes, please open an issue in https://github.com/nexB/aboutcode/, we should have a glossary there of all the terms/names we are using and have it as a reference.

Author: Thanks @AyanSinhaMahapatra and done -- see https://github.com/nexB/aboutcode/issues/139
